File: Festalytics Web App/festalytics/src/lib/firestore/bookings.js
import {
  collection,
  addDoc,
  getDocs,
  query,
  where,
  onSnapshot,
  serverTimestamp,
} from "firebase/firestore";
import { db } from "@/firebase";

const BOOKINGS_COLLECTION = "bookings";

/**
 * @param {string} docId
 * @param {object} data
 * @returns {object}
 */
export function mapBookingDocToRow(docId, data) {
  const isWalkIn =
    data.bookingSource === "walk-in" ||
    data.eventDetails?.source === "Walk-in ERP";

  const source = isWalkIn
    ? "Walk-in ERP"
    : data.bookingSource === "online" || data.eventDetails?.source === "Online Portal"
    ? "Online Portal"
    : data.bookingSource || data.eventDetails?.source || "Online Portal";

  return {
    docId,
    id: data.id || docId,
    customer: {
      name: data.customer?.name || "Client",
      email: data.customer?.contact || "No Email",
      avatar: null,
    },
    service: data.eventDetails?.category || "Wedding Event",
    bookedDate: data.bookedDate || "Today",
    eventDate: data.eventDetails?.date || "",
    timing: data.eventDetails?.timing || "",
    status: data.status || "Confirmed",
    source,
    amount: data.financials?.grandTotal || 0,
    isWalkIn,
    raw: data,
  };
}

/**
 * Persists a walk-in booking to root collection `bookings` (auto-generated doc ID).
 * Binds the venue slug on both `targetVenueId` and `eventDetails.venueId`.
 *
 * @param {string} venueSlug - e.g. "zaydan-banquet-hall"
 * @param {object} bookingPayload
 * @returns {Promise<{ bookingDocId: string }>}
 */
export async function submitWalkInBooking(venueSlug, bookingPayload) {
  if (typeof venueSlug !== "string" || !venueSlug.trim()) {
    throw new Error('submitWalkInBooking: "venueSlug" is required.');
  }
  if (!bookingPayload || typeof bookingPayload !== "object") {
    throw new Error("submitWalkInBooking: bookingPayload is required.");
  }

  try {
    const slug = venueSlug.trim();
    const record = {
      ...bookingPayload,
      targetVenueId: slug,
      eventDetails: {
        ...bookingPayload.eventDetails,
        venueId: slug,
      },
      timestamp: serverTimestamp(),
    };

    const docRef = await addDoc(collection(db, BOOKINGS_COLLECTION), record);
    return { bookingDocId: docRef.id };
  } catch (error) {
    console.error("[submitWalkInBooking] Failed to persist booking:", error);
    throw error;
  }
}

/**
 * Real-time stream of bookings for a venue (walk-in and other) by `targetVenueId`.
 *
 * @param {string} venueSlug
 * @param {(rows: object[]) => void} callback
 * @param {(error: Error) => void} [onError]
 * @returns {() => void}
 */
export function listenToVenueBookings(venueSlug, callback, onError) {
  if (typeof venueSlug !== "string" || !venueSlug.trim()) {
    const error = new Error(
      'listenToVenueBookings: "venueSlug" is required and must be a non-empty string.'
    );
    console.error("[listenToVenueBookings]", error);
    if (typeof onError === "function") onError(error);
    return () => {};
  }

  if (typeof callback !== "function") {
    throw new Error('listenToVenueBookings: "callback" must be a function.');
  }

  try {
    const bookingsQuery = query(
      collection(db, BOOKINGS_COLLECTION),
      where("targetVenueId", "==", venueSlug.trim())
    );

    return onSnapshot(
      bookingsQuery,
      (querySnapshot) => {
        const rows = querySnapshot.docs.map((d) =>
          mapBookingDocToRow(d.id, d.data())
        );
        callback(rows);
      },
      (error) => {
        console.error("[listenToVenueBookings] Snapshot listener error:", error);
        if (typeof onError === "function") onError(error);
      }
    );
  } catch (error) {
    console.error("[listenToVenueBookings] Failed to attach listener:", error);
    if (typeof onError === "function") onError(error);
    return () => {};
  }
}

/**
 * Loads older booking documents that only stored venue on eventDetails.venueId.
 * @param {string} venueSlug
 * @returns {Promise<object[]>}
 */
export async function fetchLegacyVenueBookings(venueSlug) {
  const slug = venueSlug?.trim();
  if (!slug) return [];

  try {
    const snap = await getDocs(collection(db, BOOKINGS_COLLECTION));
    const rows = [];
    snap.forEach((docSnap) => {
      const data = docSnap.data();
      const matchesVenue =
        data.targetVenueId === slug || data.eventDetails?.venueId === slug;
      if (matchesVenue && !data.targetVenueId) {
        rows.push(mapBookingDocToRow(docSnap.id, data));
      }
    });
    return rows;
  } catch (error) {
    console.error("[fetchLegacyVenueBookings] Failed to load legacy bookings:", error);
    throw error;
  }
}

File: Festalytics Mobile App/lib/web_clone/src/lib/firestore/bookings.dart
import 'dart:async';
import 'package:cloud_firestore/cloud_firestore.dart';
import '../../../../models/booking.dart';

class FirestoreBookings {
  static const String _collection = 'bookings';

  static CollectionReference<Map<String, dynamic>> get _ref =>
      FirebaseFirestore.instance.collection(_collection);

  static Map<String, dynamic> _buildWalkInRecord(
      String venueSlug, Map<String, dynamic> payload) {
    return {
      ...payload,
      'targetVenueId': venueSlug,
      'eventDetails': {
        ...(payload['eventDetails'] as Map<String, dynamic>? ?? {}),
        'venueId': venueSlug,
      },
      'timestamp': FieldValue.serverTimestamp(),
    };
  }

  static Future<String> submitWalkInBooking(
      String venueSlug, Map<String, dynamic> payload) async {
    if (venueSlug.trim().isEmpty) {
      throw ArgumentError('venueSlug is required.');
    }
    if (payload.isEmpty) {
      throw ArgumentError('bookingPayload is required.');
    }
    final record = _buildWalkInRecord(venueSlug.trim(), payload);
    final docRef = await _ref.add(record);
    return docRef.id;
  }

  static StreamSubscription<QuerySnapshot<Map<String, dynamic>>>
      listenToVenueBookings(
    String venueSlug,
    Function(List<Booking> rows) callback, {
    Function(dynamic error)? onError,
  }) {
    if (venueSlug.trim().isEmpty) {
      throw ArgumentError('venueSlug is required.');
    }
    final query = _ref.where('targetVenueId', isEqualTo: venueSlug.trim());
    return query.snapshots().listen(
      (snapshot) {
        final rows = snapshot.docs
            .map((d) => Booking.fromFirestore(
                d.id, d.data()))
            .toList();
        callback(rows);
      },
      onError: onError,
    );
  }

  static Future<List<Booking>> fetchLegacyVenueBookings(
      String venueSlug) async {
    final slug = venueSlug.trim();
    if (slug.isEmpty) return [];
    final snap = await _ref.get();
    final rows = <Booking>[];
    for (final d in snap.docs) {
      final data = d.data();
      final matchesVenue = data['targetVenueId'] == slug ||
          (data['eventDetails'] is Map &&
              (data['eventDetails'] as Map)['venueId'] == slug);
      if (matchesVenue && data['targetVenueId'] == null) {
        rows.add(Booking.fromFirestore(d.id, data));
      }
    }
    return rows;
  }

  static Future<List<Booking>> fetchAll() async {
    final snap = await _ref.get();
    return snap.docs
        .map((d) => Booking.fromFirestore(
            d.id, d.data()))
        .toList();
  }

  static Future<List<Booking>> fetchByVenue(String venueSlug) async {
    final slug = venueSlug.trim();
    if (slug.isEmpty) return [];
    final snap =
        await _ref.where('targetVenueId', isEqualTo: slug).get();
    return snap.docs
.map((d) => Booking.fromFirestore(
            d.id, d.data()))
        .toList();
}

  static Stream<QuerySnapshot<Map<String, dynamic>>> streamAll() => _ref.snapshots();

  static Future<void> updateBooking(
      String bookingId, Map<String, dynamic> data) async {
    await _ref.doc(bookingId).set(
      {...data, 'updatedAt': DateTime.now().toIso8601String()},
      SetOptions(merge: true),
    );
  }
}
